### PR TITLE
Add partners-related client functionality and test for it

### PR DESCRIPTION
**Pull request by ekoval**
https://github.com/yola/production/issues/2116

### CODE REVIEW

File: sitewit/services.py [hunk 1]
```diff
@@ -63,7 +70,7 @@ def create_account(self, site_id, url, user_name, user_email,
             data['userToken'] = user_token
 
         return self.post(
-            '/api/account/', data, headers=self._get_auth_header()).json()
+            '/api/account/', data, headers=self._get_partner_auth_header()).json()
```
Comment (snitch) [inline sitewit/services.py]: E501 line too long (82 > 79 characters)

Comment (snitch) [inline sitewit/services.py]: E501 line too long (82 > 79 characters)

File: sitewit/services.py [hunk 2]
```diff
@@ -246,7 +254,7 @@ def upgrade_campaign_subscription(self, account_token, campaign_id,
                 'currency': currency}
 
         return self.put('/api/subscription/campaign/upgrade/', data,
-                        headers=self._get_auth_header(account_token)).json()
+                        headers=self._get_account_auth_header(account_token)).json()
```
Comment (snitch) [inline sitewit/services.py]: E501 line too long (84 > 79 characters)

Comment (snitch) [inline sitewit/services.py]: E501 line too long (84 > 79 characters)

File: tests/partners/base.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        'headerTextColor': '000000',
+        'headerLogoUrl': 'https://www.partner.com/resources/sw-wl-logo.png',
+        'supportPhone': '800-555-0001',
+        'supportEmail': 'support@partner.com',
+        'supportUrl': 'https://support.partner.com',
+        'mobileAppPrimaryColor': 'cccccc',
+        'mobileAppSecondaryColor': '222222',
+        'mobileAppLogoUrl': 'https://www.partner.com/resources/sw.png',
+        'features': [
+          'SEM',
```
Comment (snitch) [inline tests/partners/base.py]: E121 continuation line under-indented for hanging indent

Comment (snitch) [inline tests/partners/base.py]: E121 continuation line under-indented for hanging indent

File: tests/partners/base.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+          'Connect',
+          'Analytics'
+        ]
+    }
+
+    partner_data = {
+      'name': uuid.uuid4(),
```
Comment (snitch) [inline tests/partners/base.py]: E121 continuation line under-indented for hanging indent

Comment (snitch) [inline tests/partners/base.py]: E121 continuation line under-indented for hanging indent

File: tests/partners/test_integration.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
+import copy
+import uuid
+
+from base import PartnerTestCase
+import sitewit.services
```
Comment (snitch) [inline tests/partners/test_integration.py]: F401 'sitewit' imported but unused

Comment (snitch) [inline tests/partners/test_integration.py]: F401 'sitewit' imported but unused

File: tests/partners/test_integration.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+    def test_partner_is_returned(self):
+        self.assertEqual(
+            deep_sort(self.create_result), deep_sort(self.get_result))
+
+
+class TestGetPartnerBadPartnerId(PartnerTestCase):
+
+    def setUp(self):
+        service = SitewitService()
```
Comment (snitch) [inline tests/partners/test_integration.py]: F841 local variable 'service' is assigned to but never used

Comment (snitch) [inline tests/partners/test_integration.py]: F841 local variable 'service' is assigned to but never used

File: tests/partners/test_integration.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+            'headerTextColor': '111111',
+            'headerLogoUrl': 'https://www.partner.com/sw.png',
+            'supportPhone': '800-555-1113',
+            'supportEmail': 'somebody@qa.com',
+            'supportUrl': 'https://support.new.com',
+            'mobileAppPrimaryColor': 'dddddd',
+            'mobileAppSecondaryColor': '333333',
+            'mobileAppLogoUrl': 'https://www.new.com/new/sw.png',
+            'features': [
+              'SEM'
```
Comment (snitch) [inline tests/partners/test_integration.py]: E121 continuation line under-indented for hanging indent

Comment (snitch) [inline tests/partners/test_integration.py]: E121 continuation line under-indented for hanging indent

File: tests/partners/test_integration.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+            'headerTextColor': '111111',
+            'headerLogoUrl': 'https://www.partner.com/sw.png',
+            'supportPhone': '800-555-1113',
+            'supportEmail': 'somebody@qa.com',
+            'supportUrl': 'https://support.new.com',
+            'mobileAppPrimaryColor': 'dddddd',
+            'mobileAppSecondaryColor': '333333',
+            'mobileAppLogoUrl': 'https://www.new.com/new/sw.png',
+            'features': [
+              'SEM', 'nonexistent feature'
```
Comment (snitch) [inline tests/partners/test_integration.py]: E121 continuation line under-indented for hanging indent

Comment (snitch) [inline tests/partners/test_integration.py]: E121 continuation line under-indented for hanging indent

**Comment by adrianmoisey** [comment]
Why did Snitch comment on these twice?

**Comment by adrianmoisey** [comment]
Why did Snitch comment on these twice?

**Comment by jouberthenk** [comment]
Can't tell if joke or github glirch

**Comment by adrianmoisey** [comment]
Mine was a joke :D 
Not sure about the bot

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

File: sitewit/services.py [hunk 9]
```diff
@@ -25,10 +25,17 @@ def __init__(self, **kwargs):
 
         super(SitewitService, self).__init__(**kwargs)
 
-    def _get_auth_header(self, account_token=None):
+    def _get_account_auth_header(self, account_token):
+        auth_raw = '%s:%s:%s' % (
+            self._partner_id, self._partner_token, account_token)
```
Comment (stefanor) [inline sitewit/services.py]:
```python
auth_raw = ':'.join((...))
``` ?

File: sitewit/services.py [hunk 10]
```diff
@@ -25,10 +25,17 @@ def __init__(self, **kwargs):
...
+
+        return {'PartnerAuth': base64.b64encode(auth_raw)}
+
+    def _get_partner_auth_header(self, subpartner_id=None):
         auth_raw = '%s:%s' % (self._partner_id, self._partner_token)
-        if account_token is not None:
-            auth_raw += ':%s' % account_token
+        if subpartner_id is not None:
+            auth_raw += ':%s' % subpartner_id
```
Comment (stefanor) [inline sitewit/services.py]: You could use the same pattern here. Building a list, and then joining with colons.

File: sitewit/services.py (hunk 9 above)
Comment (ekoval) [inline sitewit/services.py]: Yes, this is a possible option. Do you think it is better, or just another way to do the same?

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by stefanor** [comment]
He's been commenting on things twice, quite a lot.

**Comment by ekoval** [comment]
:bell:

**Comment by thomasw** [comment]
ping @RayeN, @blaix

File: sitewit/services.py [hunk 11]
```diff
@@ -25,12 +25,21 @@ def __init__(self, **kwargs):
...
         return {'PartnerAuth': base64.b64encode(auth_raw)}
 
+    def _get_partner_auth_header(self, subpartner_id=None):
+        auth_list = [self._partner_id, self._partner_token]
+        if subpartner_id is not None:
+            auth_list.append(subpartner_id)
+
+        auth_str = ':'.join(auth_list)
+
+        return {'PartnerAuth': base64.b64encode(auth_str)}
```
Comment (RayeN) [inline sitewit/services.py]: A bit of code duplication, going to remove it.

File: tests/partners/test_integration.py [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+        self.partner_name = uuid.uuid4().hex
+
+        self.result = service.create_partner(
+            self.partner_name, self.address, self.settings)
+
+        # These fields are taken from config, no need to compare them.
+        del self.result['partnerId']
+        del self.result['partnerToken']
+
+    def test_campaign_is_returned(self):
```
Comment (RayeN) [inline tests/partners/test_integration.py]: wrong test name, about to fix it.

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by RayeN** [comment]
:+1: Looks good.  We need to move on with this.

**Pull request merged**